pkg/core/statesync/neotest_test.go

```diff
@@ -22,262 +22,413 @@ func TestStateSyncModule_Init(t *testing.T) {
-		require.True(t, module.NeedHeaders())
-		require.False(t, module.NeedStorageData())
+		require.Panics(t, func() {
+			_ = module.AddMPTNodes([][]byte{})
+		})
 	})
 
-	t.Run("initialization from headers/blocks/mpt synced stages", func(t *testing.T) {
-		bcBolt, validatorsBolt, committeeBolt := chain.NewMultiWithCustomConfig(t, boltCfg)
-		eBolt := neotest.NewExecutor(t, bcBolt, validatorsBolt, committeeBolt)
-		module := bcBolt.GetStateSyncModule()
-		require.NoError(t, module.Init(bcSpout.BlockHeight()))
+	cases := []struct {
+		name string
+		cfg  func(*config.Blockchain)
+	}{
+		{"MPT-based", boltCfg},
+		{"storage-based", boltCfgStorage},
+	}
 
-		// firstly, fetch all headers to create proper DB state (where headers are in sync)
-		stateSyncPoint := (bcSpout.BlockHeight() / stateSyncInterval) * stateSyncInterval
-		var expectedHeader *block.Header
-		for i := uint32(1); i <= bcSpout.HeaderHeight(); i++ {
-			header, err := bcSpout.GetHeader(bcSpout.GetHeaderHash(i))
-			require.NoError(t, err)
-			require.NoError(t, module.AddHeaders(header))
-			if i == stateSyncPoint+1 {
-				expectedHeader = header
-			}
+	for _, cc := range cases {
+		bcSpout, validators, committee := chain.NewMultiWithCustomConfig(t, spoutCfg)
+		e := neotest.NewExecutor(t, bcSpout, validators, committee)
+		for range 2*stateSyncInterval + int(maxTraceable) + 2 {
+			e.AddNewBlock(t)
 		}
-		require.True(t, module.IsActive())
-		require.True(t, module.IsInitialized())
-		require.False(t, module.NeedHeaders())
-		require.True(t, module.NeedStorageData())
 
-		// then create new statesync module with the same DB and check that state is proper
-		// (headers are in sync)
-		module = bcBolt.GetStateSyncModule()
-		require.NoError(t, module.Init(bcSpout.BlockHeight()))
-		require.True(t, module.IsActive())
-		require.True(t, module.IsInitialized())
-		require.False(t, module.NeedHeaders())
-		require.True(t, module.NeedStorageData())
-		unknownNodes := module.GetUnknownMPTNodesBatch(2)
-		require.Equal(t, 1, len(unknownNodes))
-		require.Equal(t, expectedHeader.PrevStateRoot, unknownNodes[0])
-
-		// add a few MPT nodes to create DB state where some of MPT nodes are missing
-		count := 5
-		for {
-			unknownHashes := module.GetUnknownMPTNodesBatch(1) // restore nodes one-by-one
-			if len(unknownHashes) == 0 {
-				break
-			}
-			err := bcSpout.GetStateSyncModule().Traverse(unknownHashes[0], func(node mpt.Node, nodeBytes []byte) bool {
-				require.NoError(t, module.AddMPTNodes([][]byte{nodeBytes}))
-				return true // add nodes one-by-one
+		t.Run(cc.name, func(t *testing.T) {
+			t.Run("inactive: spout chain is too low to start state sync process", func(t *testing.T) {
+				bcBolt, _, _ := chain.NewMultiWithCustomConfig(t, cc.cfg)
+				module := bcBolt.GetStateSyncModule()
+				require.NoError(t, module.Init(uint32(2*stateSyncInterval-1)))
+				require.False(t, module.IsActive())
+				require.Equal(t, bcBolt.GetConfig(), module.GetConfig())
```

Comment:
This test is not contract storage specific, move it out of the loop, to the place where it was before.